tangelo/toolboxes/operators/operators.py

```diff
@@ -20,17 +20,102 @@
+
+            if self.n_spinorbitals != other.n_spinorbitals:
+                raise RuntimeError("n_spinorbitals must be the same for all FermionOperators.")
+            elif self.n_electrons != other.n_electrons:
+                raise RuntimeError("n_electrons must be the same for all FermionOperators.")
+            elif self.spin != other.spin:
+                raise RuntimeError("Spin must be the same for all FermionOperators.")
+
+        if isinstance(other, FermionOperator):
+            return super(FermionOperator, self).__imul__(other)
+        elif isinstance(other, of.FermionOperator):
+            if self.n_spinorbitals is not None or self.n_electrons is not None or self.spin is not None:
+                raise RuntimeError("openfermion FermionOperator did not define a necessary attribute")
+            else:
+                f_op = FermionOperator()
+                f_op.terms = other.terms.copy()
+                return super(FermionOperator, self).__imul__(f_op)
+        else:
+            return super(FermionOperator, self).__imul__(other)
+
+    def __mul__(self, other):
+        return self.__imul__(other)
+
+    def __rmul__(self, other):
+        return self.__imul__(other)
+
+    def __iadd__(self, other):
+        # Raise error if attributes are not the same across Operators.
+        if (self.n_spinorbitals is not None and other.n_spinorbitals is not None)\
+                or (self.n_electrons is not None and other.n_electrons is not None)\
+                or (self.spin is not None and other.spin is not None):
+
+            if self.n_spinorbitals != other.n_spinorbitals:
+                raise RuntimeError("n_spinorbitals must be the same for all FermionOperators.")
+            elif self.n_electrons != other.n_electrons:
+                raise RuntimeError("n_electrons must be the same for all FermionOperators.")
+            elif self.spin != other.spin:
+                raise RuntimeError("Spin must be the same for all FermionOperators.")
+
+        if isinstance(other, FermionOperator):
+            return super(FermionOperator, self).__iadd__(other)
+        elif isinstance(other, of.FermionOperator):
+            if self.n_spinorbitals is not None or self.n_electrons is not None or self.spin is not None:
+                raise RuntimeError("openfermion FermionOperator did not define a necessary attribute")
+            else:
+                f_op = FermionOperator()
+                f_op.terms = other.terms.copy()
+                return super(FermionOperator, self).__iadd__(f_op)
+
+    def __add__(self, other):
+        return self.__iadd__(other)
+
+    def __radd__(self, other):
+        return self.__iadd__(other)
+
+    def __isub__(self, other):
+        return self.__iadd__(-1. * other)
+
+    def __sub__(self, other):
+        return self.__isub__(other)
+
+    def __rsub__(self, other):
+        return self.__isub__(other)
+
+    def __eq__(self, other):
+        # Additional checks for == operator.
```

Comment:
Right now:

- Because `isinstance` will return `True` for any class that inherits from `of.FermionOperator` I make the hypothesis that when you call `f_op1 == f_op2` it actually never enters the second branch and will return `True` as long as terms match, even if there's a disagreement in spin etc. This is some of the object-oriented programming you want to get better at: because `op.FO` (for short) is a parent class of `FO`, it means any FO instance (object)is also an instance of `op.FO`. So `isinstance` returns `True` for both.

Knowing that we can be be crafty. I think this code block can become your entire function

```
if isinstance(other,  FermionOperator) : # Only works if it's a Tangelo FO, or any future child classes of it.
   a = super(FermionOperator, self).__eq__(super(FermionOperator, other))
   b = (self.n_spinorbitals, self.n_electrons, self.spin) == (other.n_spinorbitals, other.n_electrons, other.spin)
   return a and b
elif isinstance(other,  of.FermionOperator)  # Can't be a Tangelo FO, otherwise we got into the 1st branch
   return super(FermionOperator, self).__eq__(other)
else: # Should raise an error
    Raise TypeError(f"Cannot compare objects of type {self.__class__} and {other.__class__}")
```